alexandria/core/filters.py

```diff
@@ -168,45 +169,89 @@ class FileFilterSet(FilterSet):
+                    )
+                    .exclude(pk=OuterRef("pk")),
+                )
+            )
+        return qs
 
     class Meta:
         model = models.File
-        fields = ["original", "renderings", "variant", "metainfo", "files"]
+        fields = [
+            "original",
+            "renderings",
+            "variant",
+            "metainfo",
+            "files",
+            "only_newest",
+        ]
 
 
-class FileSearchFilterSet(FileFilterSet):
+class SearchFilterSet(FileFilterSet):
     query = CharFilter(method="search_files")
 
+    # Cut-off value, we don't want all the garbage matches
+    min_rank = NumberFilter(field_name="search_rank", lookup_expr="gte")
+
+    def filter_queryset(self, queryset):
+        """Enforce some defaults in the queryset.
+
+        First, we only ever want to search originals, so no thumbnails
+        shall be searched. Secondly, we don't return any results if
+        the user didn't search for anything.
+        """
+        queryset = queryset.filter(
+            # We only ever search originals
+            variant=models.File.Variant.ORIGINAL
+        )
+
+        if not self.form.cleaned_data.get("query"):
+            # If no search parameter is given, we do not return any value
+            return queryset.none()
+        return super().filter_queryset(queryset)
+
     def search_files(self, queryset, name, value):
         search_queries = [
-            Q(
-                content_vector=SearchQuery(
-                    value,
-                    config="simple",
-                    search_type=settings.ALEXANDRIA_CONTENT_SEARCH_TYPE,
-                )
+            SearchQuery(
+                value,
+                config="simple",
+                search_type=settings.ALEXANDRIA_CONTENT_SEARCH_TYPE,
             )
         ]
         for lang, config in settings.ALEXANDRIA_ISO_639_TO_PSQL_SEARCH_CONFIG.items():
-            search_queries.append(
-                Q(
-                    content_vector=SearchQuery(
-                        value,
-                        config=config,
-                        search_type=settings.ALEXANDRIA_CONTENT_SEARCH_TYPE,
-                    )
-                )
+            query_fragment = SearchQuery(
+                value,
+                config=config,
+                search_type=settings.ALEXANDRIA_CONTENT_SEARCH_TYPE,
             )
+            search_queries.append(query_fragment)
 
         search_query = reduce(or_, search_queries)
 
-        latest_files = models.File.objects.filter(
-            document=OuterRef("document_id"), variant=models.File.Variant.ORIGINAL
-        ).order_by("-created_at")
-        queryset = queryset.filter(pk=Subquery(latest_files.values("pk")[:1])).filter(
-            search_query
+        queryset = queryset.annotate(
+            search_rank=SearchRank(F("content_vector"), search_query),
+            search_context=SearchHeadline(F("content_text"), search_query),
         )
```

Comment:
Note: It's possible that we'll need to do a `.filter(content_vector=search_query)` as well to improve performance (instead of ordering by `search_rank`, which "only" sorts the best matches first)

Reply:
Updated the code, we're now doing a `.filter()` call